tests/integration/devfile/cmd_devfile_status_test.go

```diff
@@ -38,325 +38,301 @@ var _ = Describe("odo devfile status command tests", func() {
-					urlParams = append(urlParams, "--host", urlHost)
-				}
+					statusEntry := utils.GetMostRecentEventOfType(machineoutput.TypeKubernetesPodStatus, entries, false)
 
-				helper.Cmd("odo", urlParams...).ShouldPass()
+					if statusEntry == nil {
+						return false
+					}
 
-				helper.Cmd("odo", "push", "-o", "json", "--project", namespace).ShouldPass()
+					podStatus := statusEntry.(*machineoutput.KubernetesPodStatus)
 
-				session := helper.CmdRunner("odo", "component", "status", "-o", "json", "--project", namespace, "--follow")
+					// Check if a pod is running and correct, and different from the old pod; returns "" if success, otherwise returns the reason why not
+					checkPod := func(pod machineoutput.KubernetesPodStatusEntry) string {
 
-				helper.WaitForOutputToContain("urlReachable", 180, 10, session)
+						if pod.Name == oldPodName {
+							return "Skipping old pod"
+						}
 
-				stdoutContents := string(session.Out.Contents())
+						if pod.Phase != "Running" {
+							return "Phase is not running"
+						}
 
-				entries, err := utils.ParseMachineEventJSONLines(stdoutContents)
-				Expect(err).NotTo(HaveOccurred())
+						match := false
+						for _, labelValue := range pod.Labels {
+							if labelValue == cmpName {
+								match = true
+								break
+							}
+						}
+						if !match {
+							return "No matching labels"
+						}
 
-				// Verify url status is present and correct
-				urlReachableEntry := utils.GetMostRecentEventOfType(machineoutput.TypeURLReachable, entries, true).(*machineoutput.URLReachable)
+						match = false
+						for _, container := range pod.Containers {
+							if container.Name == "runtime" && container.State.Running != nil {
+								match = true
+								break
+							}
+						}
+						if !match {
+							return "Could not find runtime container"
+						}
+						return ""
+					} // end checkPod
 
-				expectedKind := "ingress"
-				if !ingress || openshift {
-					expectedKind = "route"
-				}
+					for _, pod := range podStatus.Pods {
+						failReason := checkPod(pod)
+						if failReason == "" {
+							return true
+						}
+						fmt.Println("Pod", pod.Name, "did not satisfy condition:", failReason)
+					}
 
-				Expect(urlReachableEntry.Kind).To(Equal(expectedKind))
-				Expect(urlReachableEntry.Reachable).To(Equal(!ingress || openshift)) // On non-openshift, the ingress URL is using a random hostname, so should not be resolveable
-				Expect(urlReachableEntry.Port).To(Equal(3000))
-				// Expect(urlReachableEntry.Secure).To(Equal(secure))
+					return false
+				}, 180, 10).Should(Equal(true))
 
 				utils.TerminateSession(session)
 
-			})
+			}) // end It
+		})
+	})
+
+	Context("Verify URL status is correctly reported", func() {
 
-		}
+		testCombo(false, false, "Route Nonsecure")
+		testCombo(false, true, "Route Secure")
+		testCombo(true, false, "Ingress Nonsecure")
+		testCombo(true, true, "Ingress Secure")
 
 	}) // end context
```

Comment:
I think this can be written better with the `When` above.
```go
var _ = Describe("...", func() {
    	testCombo := func(ingress bool, route bool, name string) {
		defer GinkgoRecover()
		It("Verify that odo component status detects the URL status: "+name, func() {
			openshift := os.Getenv("KUBERNETES") != "true"
			if !ingress && !openshift {
				Skip("Route-based URLs is an OpenShift only scenario")
			}
			urlHost := helper.RandString(12) + ".com"

			urlParams := []string{"url", "create", "my-url", "--port", "3000"}
			if !strings.Contains(name, "Nonsecure") {
				urlParams = append(urlParams, "--secure")
			}

			if ingress {
				urlParams = append(urlParams, "--ingress")
				urlParams = append(urlParams, "--host", urlHost)
			}

			helper.Cmd("odo", urlParams...).ShouldPass()

			helper.Cmd("odo", "push", "-o", "json", "--project", commonVar.Project).ShouldPass()

			session := helper.CmdRunner("odo", "component", "status", "-o", "json", "--project", commonVar.Project, "--follow")

			helper.WaitForOutputToContain("urlReachable", 180, 10, session)

			stdoutContents := string(session.Out.Contents())

			entries, err := utils.ParseMachineEventJSONLines(stdoutContents)
			Expect(err).NotTo(HaveOccurred())

			// Verify url status is present and correct
			urlReachableEntry := utils.GetMostRecentEventOfType(machineoutput.TypeURLReachable, entries, true).(*machineoutput.URLReachable)

			expectedKind := "ingress"
			if !ingress || openshift {
				expectedKind = "route"
			}

			Expect(urlReachableEntry.Kind).To(Equal(expectedKind))
			Expect(urlReachableEntry.Reachable).To(Equal(!ingress || openshift)) // On non-openshift, the ingress URL is using a random hostname, so should not be resolveable
			Expect(urlReachableEntry.Port).To(Equal(3000))
			if !strings.Contains(name, "Nonsecure") {
				Expect(urlReachableEntry.Secure).To(Equal("Secure"))
			}

			utils.TerminateSession(session)

		})
	}

    When("a component is created with --devfile", func(){
    BeforeEach(...)

    testCombo(false, false, "Route Nonsecure")
    testCombo(false, false, "Route Nonsecure")
    testCombo(true, false, "Ingress Nonsecure")
    testCombo(true, true, "Ingress Secure")

    When("doing odo push", func() {...}
})
```
The above will still create a component for the testCombo cases, but we can let that go for better structure in my opinion.

Reply:
I tried but it complained about where I can use BeforeEach.

Reply:
It seems to be working fine for me. Check this - 
```go
var _ = Describe("odo devfile status command tests", func() {
	...
	var _ = BeforeEach(func() {...})

	// Clean up after the test
	// This is run after every Spec (It)
	var _ = AfterEach(func() {...})

	//Function used to test context: "Verify URL status is correctly reported"
	testCombo := func(ingress bool, secure bool, name string) {
		defer GinkgoRecover()
		It("Verify that odo component status detects the URL status: "+name, func() {...})
	}

	When("Creating nodejs component using devfile", func() {
		BeforeEach(func() {
			helper.Cmd("odo", "create", "--project", namespace, cmpName, "--devfile", helper.GetExamplePath("source", "devfiles", "nodejs", "devfile.yaml")).ShouldPass()
			helper.CopyExample(filepath.Join("source", "devfiles", "nodejs", "project"), context)
		})
		testCombo(false, false, "Route Nonsecure")
		testCombo(true, false, "Ingress Nonsecure")
		// testCombo(false, true, "Route Secure")   # Commented until issue https://github.com/openshift/odo/issues/5217 gets fixed
		// testCombo(true, true, "Ingress Secure")  # Commented until issue https://github.com/openshift/odo/issues/5217 gets fixed

		When("doing odo push", func() {
			BeforeEach(func() {
				helper.Cmd("odo", "push", "-o", "json", "--project", namespace).ShouldPass()
			})
			It("Verify that odo component status correctly reports supervisord status", func() {...})

			It("Verify that odo component status correctly detects component Kubernetes pods", func() {...}) // end It
		})
	})

})
```